lib/internal/modules/esm/resolve.js

```diff
@@ -1,26 +1,604 @@
+      if (packageConfig === packageConfigCandidate) {
+        packageJSONUrl = pathToFileURL(filename);
+        break;
+      }
+    }
+    if (packageJSONUrl !== undefined &&
+        packageConfig.name === packageName &&
+        packageConfig.exports !== undefined) {
+      if (packageSubpath === './') {
+        return new URL('./', packageJSONUrl);
+      } else if (packageSubpath === '') {
+        return packageMainResolve(packageJSONUrl, packageConfig, base);
+      } else {
+        return packageExportsResolve(
+          packageJSONUrl, packageSubpath, packageConfig, base);
+      }
+    }
+  }
+
+  let packageJSONUrl =
+    new URL('./node_modules/' + packageName + '/package.json', base);
+  let packageJSONPath = fileURLToPath(packageJSONUrl);
+  let lastPath;
+  do {
+    const stat = tryStatSync(
+      StringPrototypeSubstr(packageJSONPath, 0, packageJSONPath.length - 13));
+    if (!stat.isDirectory()) {
+      lastPath = packageJSONPath;
+      packageJSONUrl = new URL((isScoped ?
+        '../../../../node_modules/' : '../../../node_modules/') +
+        packageName + '/package.json', packageJSONUrl);
+      packageJSONPath = fileURLToPath(packageJSONUrl);
+      continue;
+    }
+
+    // Package match.
+    const packageConfig = getPackageConfig(packageJSONPath, base);
+    if (packageSubpath === './') {
+      return new URL('./', packageJSONUrl);
+    } else if (packageSubpath === '') {
+      return packageMainResolve(packageJSONUrl, packageConfig, base);
+    } else if (packageConfig.exports !== undefined) {
+      return packageExportsResolve(
+        packageJSONUrl, packageSubpath, packageConfig, base);
+    } else {
+      return finalizeResolution(
+        new URL(packageSubpath, packageJSONUrl), base);
+    }
+    // Cross-platform root check.
+  } while (packageJSONPath.length !== lastPath.length);
+
+  // eslint can't handle the above code.
+  // eslint-disable-next-line no-unreachable
+  throw new ERR_MODULE_NOT_FOUND(packageName, fileURLToPath(base));
+}
+
+function shouldBeTreatedAsRelativeOrAbsolutePath(specifier) {
+  if (specifier === '') return false;
+  if (specifier[0] === '/') return true;
+  if (specifier[0] === '.') {
+    if (specifier.length === 1 || specifier[1] === '/') return true;
+    if (specifier[1] === '.') {
+      if (specifier.length === 2 || specifier[2] === '/') return true;
+    }
+  }
+  return false;
+}
+
+function moduleResolve(specifier /* string */, base /* URL */) { /* -> URL */
+  // Order swapped from spec for minor perf gain.
+  // Ok since relative URLs cannot parse as URLs.
+  let resolved;
+  if (shouldBeTreatedAsRelativeOrAbsolutePath(specifier)) {
+    resolved = new URL(specifier, base);
+  } else {
+    try {
+      resolved = new URL(specifier);
+    } catch {
+      return packageResolve(specifier, base);
+    }
+  }
+  return finalizeResolution(resolved, base);
+}
+
+function defaultResolve(specifier, { parentURL } = {}) {
```

Comment:
I think we should keep the `defaultResolve` third parameter, as this function is what the hook is replicating: https://nodejs.org/api/esm.html#esm_code_resolve_code_hook. I know it's unused.

Ditto for `defaultGetFormat` and the other `default*` parameters.

Reply:
I'd like to see these arguments removed because they do very weird things here: They shadow the function name. Since these are the very last functions in the chain and will never call any others, accepting `defaultResolve` as an argument is confusing.

Reply:
@jkrems can we do that in a separate PR?

Reply:
@jkrems Fwiw, I dislike this for the same reason – I’ve renamed them so that they do not shadow the function itself anymore + so that it’s clear that they are unused